File: 102_Combinations.cpp
// Combinations
// Return all combinations of k numbers from 1..n.
// Solution: backtracking.
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> combine(int n, int k){
    vector<vector<int>> res; vector<int> cur;
    function<void(int)> dfs = [&](int start){
        if(cur.size()==k){ res.push_back(cur); return; }
        for(int i=start;i<=n;++i){ cur.push_back(i); dfs(i+1); cur.pop_back(); }
    };
    dfs(1);
    return res;
}

int main(){
    auto r = combine(4,2);
    cout<<"Combinations:\n"; for(auto &v:r){ for(int x:v) cout<<x; cout<<"\n"; }
    return 0;
}

// Time Complexity: O(C(n,k) * k)
// Space Complexity: O(k)

File: 102_Combinations.py
"""
Combinations
Return all combinations of k numbers chosen from 1..n.
"""
from typing import List

def combine(n: int, k: int) -> List[List[int]]:
    res = []
    def dfs(start, cur):
        if len(cur) == k:
            res.append(cur.copy()); return
        for i in range(start, n+1):
            cur.append(i); dfs(i+1, cur); cur.pop()
    dfs(1, [])
    return res

if __name__ == '__main__':
    print('Combinations:', combine(4,2))

# Time Complexity: O(C(n,k) * k)
# Space Complexity: O(k)
